rust/lance-linalg/src/kmeans.rs

```diff
@@ -329,21 +297,28 @@ impl KMeans {
             ));
         }
 
-        if !matches!(data.value_type(), DataType::Float32) {
+        if !data.value_type().is_floating() {
             return Err(ArrowError::InvalidArgumentError(format!(
-                "KMeans: data must be Float32, got: {}",
+                "KMeans: data must be floating number, got: {}",
                 data.value_type()
             )));
         }
-        let values: &Float32Array = data.values().as_primitive();
 
+        let data = data
+            .values()
+            .as_any()
+            .downcast_ref::<T::ArrayType>()
+            .ok_or(Error::InvalidArgumentError(format!(
+                "KMeans: data must be floating number, got: {}",
+                data.value_type()
+            )))?;
```

Comment:
Would this work?
```suggestion
        let data = as_primitive_array::<T>(data);
```
(requires `use arrow_array::cast::as_primitive_array`)

Reply:
It wont work because we wanted to support https://github.com/lancedb/lance/blob/main/rust/lance-arrow/src/bfloat16.rs#L31 , which is not a PrimiritiveArray today.

It'd be great to have that in arrow-rust one day tho.

Reply:
Oh, got it.